src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -326,11 +382,13 @@ public void OnResponseAbort(Exception abortException)
                     {
                         throw new IOException(SR.net_http_request_aborted, _abortException);
                     }
-                    else if (_state == StreamState.ExpectingHeaders)
+                    else if (_state == StreamState.ExpectingHeaders || _state == StreamState.ExpectingIgnoredHeaders || _state == StreamState.ExpectingStatus)
                     {
-                        Debug.Assert(!_hasWaiter);
-                        _hasWaiter = true;
-                        _waitSource.Reset();
+                        if (!_hasWaiter)
```

Comment:
Why this change? What makes it so we can get into the case where _hasWaiter is true, when it couldn't be before?

Reply:
What is exact purpose of TryEnsureHeaders() @geoffkizer ? It seems like original logic would check if we have complete headers and if not it would wait once and assume we will never go back and we would proceed with processing body. 

Now with 1xx when we get 100 we get full header set but we will loop and expect another set of header block. I was hitting the assert before as well and that was reason why I would  move back with state instead of adding new state as you suggested.

Reply:
If _hasWaiter is true, that means we currently have someone waiting for the headers. That shouldn't be the case here. 

Most likely, I think that we are just not resetting _hasWaiter to false, but I'm not sure how that happens.

Either way, it seems like we shouldn't remove the assert and change the logic here.

Reply:
I'll take another look. I did try to reset _hasWaiter in  OnResponseHeadersComplete() when we finish processing 1xx but that was not sufficient for some reason.

Reply:
The sequence is like this @geoffkizer:
 - we call TryEnsureHeaders() and set awaiter as don't have headers. 
- we wait process headers and call TryEnsureHeaders() again. 
- for 1xx we are in state ExpectingStatus. this is just like when we started. So that creates awaiter. 
- we do 100 processing and loop back. 
- now we call TryEnsureHeaders() again with awaiter set.

We could grab lock and reset that somewhere ReadResponseHeadersAsync() or modify  TryEnsureHeaders() to be smarted and not only look at state but also Response.StatusCode.

Let me know what is your preference.

Reply:
I think there's a timing issue here. 

The On* routines (e.g. OnResponseHeader, OnResponseHeadersComplete) get called asynchronously as we process received frames off the connection. They need to be able to update the state of the Http2Stream and response object, and then signal the waiting client task if necessary. We can't control when the client task will actually process the received data, so we need to store all relevant state on the Http2Stream or the response object. The waiter mechanism basically just tells the client task "more data has arrived, so wake up and try to process it".

The problem here is that we used to assume that we only received one header block per response (not counting trailing headers). But now, we can receive multiple header blocks before we receive the body. The logic below in ReadResponseHeadersAsync is assuming that it can process each header block individually. But that's *not* the case, and we shouldn't try to make it the case.

For example:
(1) ReadResponseHeadersAsync calls TryEnsureHeaders and has to wait because headers are available
(2) We receive a header block with status = 100, process it, eventually OnResponseHeadersComplete is called and the waiter will be signalled, but *not necessarily executed immediately*
(3) Now, we receive another header block. Status could be 200 or 103 or 500 or whatever. This will overwrite the existing status on the response object. That's bad.

I think we should change the approach here. TryEnsureHeaders should not signal the waiter for a 1xx response. It should only signal the waiter when it receives the final response headers. This maintains the approach today that TryEnsureHeaders only happens once.

Additionally, we should store the Task for the 100-continue notification on the Http2Stream object, so that when OnResponseHeader receives a 100-continue (or other) status, it can signal the task directly from that code. No need to signal the waiter here, just signal the task directly.

Thoughts?

Reply:
BTW, I'm not quite sure how this causes _hasWaiter to be true here, but it seems related.